### ISSUE TITLE
Bug: YAML comment is removed

### ISSUE THREAD

**Issue by karmeye** [platers/obsidian-linter#660]
## Describe the Bug

A YAML comment on its own line is removed when linting. I can't find any rule that does this.

## How to Reproduce

```yaml
created: 2023-02-18T10:53:01+00:00
# Dont remove me
disabled rules: [capitalize-headings]
modified: 2023-03-20T14:53:42+00:00

```

## Expected Behavior

YAML comnents should not be removed.

## Device

- [probably] Desktop
- [x] Mobile

**Comment by pjkaufman** [comment]
Thanks for reporting this @karmeye . Could you provide your `data.json` so I can try to reproduce this. 

The Linter is not able to garuantee that comments will not be removed, but it will try to preserve them as best it can.

**Comment by pjkaufman** [comment]
I am taking a closer look at the logic in question, but I am not sure we will be able to accommodate comments in the YAML. I am taking a look at whether or not it is feasible to preserve them, but it looks like sort is not able to preserve the comments. I am still looking, so it may still be possible in a way I have not yet considered, but so far it is looking like the removal comments is not something that is easily preventable in this case.

**Comment by pjkaufman** [comment]
> So it's only an issue when key sort is enabled?
> 
> It's quite serious though, isn't it? Users whi don't know about the issue might bulk lint a folder and unknowingly lose a lot of potentially important comments.
> 
> I have no idea why it's complicated though... I appreciate your work +1

This does seem to be limited to the key sort, but other operations can remove them as well.

The reason it is complicated to keep a comment in YAML is it is hard to tell what is a comment accurately and what it belongs to. I am not sure I have seen a good parser that would allow for this which means using an external library like is used in the Linter will not preserve the comments by default. Thus custom logic has to be added for accounting for comments. In this case, the reason it is so difficult is that sort is designed to grab the key value and ignore other values to help move them around. Exactly how this is getting rid of the comment, I am not sure yet, but it seems to be related to copying the value of a key from one place to another in the YAML.

**Comment by pjkaufman** [comment]
Hopefully I will have time to take a closer look at this this weekend.

**Comment by breakid** [comment]
There is a Visual Studio Code extension called [YAML Sort](https://github.com/pascalre/vscode-yaml-sort) that can maintain comments. Would any of the logic from that help address this issue?

**Comment by pjkaufman** [comment]
The logic that is used to sort the keys is not feasible to use unless we cut support for a lot of customization in the YAML based on what that extension does [here](https://github.com/pascalre/vscode-yaml-sort/blob/2b4d31c19538dcafdd5ad353799954bb66d59d9c/src/adapter/js-yaml-adapter.ts#L40-L64). Several settings are passed in to the `js-yaml` object, but then it formats a bunch of different places. Since the Linter allows some level of customization that is not feasible if we let `js-yaml` do the sorting, we could use it as a reference for how things would be sorted, but we would not be able to use it as the actual thing that outputs the sorted results. This would likely still have most if not all of the same problems of the current sort logic not respecting the line continuation indicators and the comments.

In summary, it does give an idea of one way to proceed, it may not actually fix anything with this problem.

**Comment by pjkaufman** [comment]
I have merged a change that should fix this. It may cause other issues, so I plan to put this out as beta release that will hopefully help us determine any issues that exist prior to having it roll out to all users.
